src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -368,78 +378,76 @@ public synchronized void removeShard(int shardId, String reason) throws Elastics
+                shardInjector.getInstance(Store.class).close();
             } catch (Throwable e) {
-                logger.debug("[{}] failed to close translog", e, shardId);
-                // ignore
+                logger.warn("[{}] failed to close store on shard removal (reason: [{}])", e, shardId, reason);
             }
+        }
+    }
+
+    /**
+     * This method gets an instance for each of the given classes passed and calls #close() on the returned instance.
+     * NOTE: this method swallows all exceptions thrown from the close method of the injector and logs them as debug log
+     */
+    private void closeInjectorResource(ShardId shardId, Injector shardInjector, Class<? extends Closeable>... toClose) {
+        for (Class<? extends Closeable> closeable : toClose) {
             try {
-                // now we can close the translog
-                shardInjector.getInstance(PercolatorQueriesRegistry.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close PercolatorQueriesRegistry", e, shardId);
-                // ignore
+                final Closeable instance = shardInjector.getInstance(closeable);
+                if (instance == null) {
+                    throw new NullPointerException("No instance available for " + closeable.getName());
+                }
+                IOUtils.close(instance);
+            } catch (Throwable t) {
+                logger.debug("{} failed to close {}", t, shardId, Strings.toUnderscoreCase(closeable.getSimpleName()));
             }
+        }
+    }
 
-            // call this before we close the store, so we can release resources for it
-            indicesLifecycle.afterIndexShardClosed(sId, indexShard);
-            // if we delete or have no gateway or the store is not persistent, clean the store...
-            final Store store = shardInjector.getInstance(Store.class);
-            // and close it
+    private void onShardClose(ShardLock lock) {
+        if (deleted.get()) { // we remove that shards content if this index has been deleted
             try {
-                    store.close();
-            } catch (Throwable e) {
-                logger.warn("[{}] failed to close store on shard deletion", e, shardId);
+                nodeEnv.deleteShardDirectorySafe(lock, indexSettings);
+            } catch (IOException e) {
+                logger.warn("{} failed to delete shard content", e, lock.getShardId());
             }
-            Injectors.close(injector);
+        }
+    }
+
+    private class StoreCloseListener implements Store.OnClose {
+        private final ShardId shardId;
 
-            logger.debug("[{}] closed (reason: [{}])", shardId, reason);
-        } catch (Throwable t) {
-            throw t;
+        public StoreCloseListener(ShardId shardId) {
+            this.shardId = shardId;
         }
 
+        @Override
+        public void handle(ShardLock lock) {
```

Comment:
can we add a comment/doc that this is called under the shard lock? will also be a good idea to assert on it.